jobmaster/dm/config/config.go

```diff
@@ -0,0 +1,200 @@
+	// EnableHeartbeat bool `yaml:"enable-heartbeat" toml:"enable-heartbeat" json:"enable-heartbeat"`
+	// HeartbeatUpdateInterval int `yaml:"heartbeat-update-interval" toml:"heartbeat-update-interval" json:"heartbeat-update-interval"`
+	// HeartbeatReportInterval int    `yaml:"heartbeat-report-interval" toml:"heartbeat-report-interval" json:"heartbeat-report-interval"`
+	// pt/gh-ost name rule,support regex
+	// OnlineDDLScheme string `yaml:"online-ddl-scheme" toml:"online-ddl-scheme" json:"online-ddl-scheme"`
+	// BWList map[string]*filter.Rules `yaml:"black-white-list" toml:"black-white-list" json:"black-white-list"`
+	// EnableANSIQuotes bool `yaml:"ansi-quotes" toml:"ansi-quotes" json:"ansi-quotes"`
+	// RemoveMeta bool `yaml:"remove-meta"`
+}
+
+// The difference between task configuration and job configuration is that a task has only one usptream.
+type TaskCfg JobCfg
+
+// UpstreamCfg add db-config to dmconfig.MySQLInstance, because we no need source cfg now.
+type UpstreamCfg struct {
+	dmconfig.MySQLInstance `yaml:",inline" toml:",inline" json:",inline"`
+	DBCfg                  *dmconfig.DBConfig `yaml:"db-config" toml:"db-config" json:"db-config"`
+}
+
+func (c *JobCfg) DecodeFile(fpath string) error {
+	bs, err := os.ReadFile(fpath)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if err = yaml.UnmarshalStrict(bs, c); err != nil {
+		return err
+	}
+	return c.adjust()
+}
+
+// copy from tiflow/dm/config/config.go#adjust
+// should be refactor
+func (c *JobCfg) adjust() error {
+	for _, upstream := range c.Upstreams {
+		if len(upstream.MydumperConfigName) > 0 {
+			rule, ok := c.Mydumpers[upstream.MydumperConfigName]
+			if !ok {
+				return errors.Errorf("mydumper config %s not exist in mydumpers", upstream.MydumperConfigName)
+			}
+			upstream.Mydumper = new(dmconfig.MydumperConfig)
+			*upstream.Mydumper = *rule
+		}
+		if len(upstream.LoaderConfigName) > 0 {
+			rule, ok := c.Loaders[upstream.LoaderConfigName]
+			if !ok {
+				return errors.Errorf("loader config %s not exist in loaders", upstream.LoaderConfigName)
+			}
+			upstream.Loader = new(dmconfig.LoaderConfig)
+			*upstream.Loader = *rule // ref loader config
+		}
+		if len(upstream.SyncerConfigName) > 0 {
+			rule, ok := c.Syncers[upstream.SyncerConfigName]
+			if !ok {
+				return errors.Errorf("syncer config %s not exist in syncers", upstream.SyncerConfigName)
+			}
+			upstream.Syncer = new(dmconfig.SyncerConfig)
+			*upstream.Syncer = *rule // ref syncer config
+		}
+	}
+
+	return nil
+}
+
+func (c *JobCfg) Yaml() (string, error) {
+	b, err := yaml.Marshal(c)
+	return string(b), err
+}
+
+func (c *JobCfg) Clone() (*JobCfg, error) {
+	content, err := c.Yaml()
+	if err != nil {
+		return nil, err
+	}
+	clone := &JobCfg{}
+	err = yaml.Unmarshal([]byte(content), clone)
+	return clone, err
+}
+
+// Generate task-id by host:port
+func GenerateTaskID(upstreamCfg *UpstreamCfg) string {
+	return upstreamCfg.DBCfg.Host + ":" + strconv.Itoa(upstreamCfg.DBCfg.Port)
```

Comment:
user name and password also matter. In other words, I think DSN is an ID

Reply:
Why, I think host and port can determine an upstream

Reply:
I add source-id in [751d477](https://github.com/hanfei1991/microcosm/pull/227/commits/751d4772c44ae62f3a29af1326fb3ef7967c2f3f), I think we still need it, for some one who may change host for a upstream.